<p><font color="blue">
    <h3 align="center">Laboratoire d'Etudes Statistiques et de Conception d'Applications et Logiciels</h3>
    <h1 align="center">L.E.S.C.A.L.</h3>
    <h1 align="center">--------------------------------------</h3>
    <h2 align="center">Cedoriase Academy For Artificial Intelligence</h2>
    <h4 align="center">+229 66233516 / cedoriaseacademy@lescal-soc.com</h4>
    <h1 align="center">--------------------------------------</h3>
    <h2 align="center">Analyse exploratoire de données - Partie 1</h2>
    <h1 align="center">--------------------------------------</h3>
</font></p>

#### Explorer les données collectées auprès d'une population (clients d'une entreprises, habitants d'un quartier) permet de comprendre les habitudes et comportements de cette population. L'analyse exploratoire des données est aussi qualifiée de fouille de données ou data mining

#### Sujet : Nous exposerons les principales méthodes de la librairie pandas en analysant un ensemble de données sur le taux de résiliation des clients des opérateurs télécoms

### 1 Importation et affichage des données

In [43]:
import pandas as pd
df= pd.read_csv("telecom_churn.csv", delimiter= ";")
df.head(60)

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
0,KS,128,415,No,Yes,25,265.1,110,45.07,197.4,99,16.78,244.7,91,11.01,10.0,3,2.70,1,False
1,OH,107,415,No,Yes,26,161.6,123,27.47,195.5,103,16.62,254.4,103,11.45,13.7,3,3.70,1,False
2,NJ,137,415,No,No,0,243.4,114,41.38,121.2,110,10.30,162.6,104,7.32,12.2,5,3.29,0,False
3,OH,84,408,Yes,No,0,299.4,71,50.90,61.9,88,5.26,196.9,89,8.86,6.6,7,1.78,2,False
4,OK,75,415,Yes,No,0,166.7,113,28.34,148.3,122,12.61,186.9,121,8.41,10.1,3,2.73,3,False
5,AL,118,510,Yes,No,0,223.4,98,37.98,220.6,101,18.75,203.9,118,9.18,6.3,6,1.70,0,False
6,MA,121,510,No,Yes,24,218.2,88,37.09,348.5,108,29.62,212.6,118,9.57,7.5,7,2.03,3,False
7,MO,147,415,Yes,No,0,157.0,79,26.69,103.1,94,8.76,211.8,96,9.53,7.1,6,1.92,0,False
8,LA,117,408,No,No,0,184.5,97,31.37,351.6,80,29.89,215.8,90,9.71,8.7,4,2.35,1,False
9,WV,141,415,Yes,Yes,37,258.6,84,43.96,222.0,111,18.87,326.4,97,14.69,11.2,5,3.02,0,False


In [44]:
# dimensions du tableau 
print(df.shape)

(3333, 20)


In [45]:
# affichage des noms des colonnes
print(df.columns)

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')


In [46]:
# obtenir des infos générales sur le tableau de données
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

- bool, int64, float64 et object sont les types de données de nos variables. 

- On voit qu'une variable est de type logique (bool), 

- 3 variables sont de type object, 

- et 16 variables sont numériques. 

- int & float64 sont des "variables numérique"

- object est une "variable non numérique"

Avec cette même méthode, nous pouvons facilement voir s'il y a des valeurs manquantes. Ici, il n'y en a pas car chaque colonne contient 3333 observations, le même nombre de lignes que nous avons vu précédemment avec shape.

### 2 Statistiques descriptives des variables quantitatives
 
La méthode describe() donne les statistiques descriptives de base de chaque variable numérique (types int64 et float64) : nombre de valeurs non manquantes, moyenne, écart type, médiane, quartiles 0,25 et 0,75.

In [47]:
df.describe()

,Account length,Area code,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls
count,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000
mean,101.064806,437.182418,8.099010,179.775098,100.435644,30.562307,200.980348,100.114311,17.083540,200.872037,100.107711,9.039325,10.237294,4.479448,2.764581,1.562856
std,39.822106,42.371290,13.688365,54.467389,20.069084,9.259435,50.713844,19.922625,4.310668,50.573847,19.568609,2.275873,2.791840,2.461214,0.753773,1.315491
min,1.000000,408.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,23.200000,33.000000,1.040000,0.000000,0.000000,0.000000,0.000000
25%,74.000000,408.000000,0.000000,143.700000,87.000000,24.430000,166.600000,87.000000,14.160000,167.000000,87.000000,7.520000,8.500000,3.000000,2.300000,1.000000
50%,101.000000,415.000000,0.000000,179.400000,101.000000,30.500000,201.400000,100.000000,17.120000,201.200000,100.000000,9.050000,10.300000,4.000000,2.780000,1.000000
75%,127.000000,510.000000,20.000000,216.400000,114.000000,36.790000,235.300000,114.000000,20.000000,235.300000,113.000000,10.590000,12.100000,6.000000,3.270000,2.000000
max,243.000000,510.000000,51.000000,350.800000,165.000000,59.640000,363.700000,170.000000,30.910000,395.000000,175.000000,17.770000,20.000000,20.000000,5.400000,9.000000


##### Interprétation de quelques quartiles

- Variable : Account length

1er quartile (25%) : 25% des abonnés disposent d'un numéro dont le nombre de chiffres est inférieur à 74

- Variable :Total night minutes

2è quartile ou médiane (50%) : 50% des abonnés effectuent moins de 201,2 minutes d'appel dans la nuit

### 3 Statistiques descriptives des variables qualitatives

Pour voir les statistiques sur les variables non numériques (qualitatives), il faut indiquer explicitement les types de données d'intérêt dans le paramètre include.

In [48]:
df.describe(include= ["object", "bool"])

,State,International plan,Voice mail plan,Churn
count,3333,3333,3333,3333
unique,51,2,2,2
top,WV,No,No,False
freq,106,3010,2411,2850


##### Interprétation State

- count : La variable State comporte 3333 observations non manquantes

- unique : Elle comporte 51 observations uniques

- top (Modalité la plus fréquente) : L'état dont provient la majorité des abonnés est WV

- freq : L'effectif associé à la modalité la plus fréquente est 106

Pour les variables qualitatives (type object) et booléennes (type bool), nous pouvons utiliser la méthode value_counts. Regardons la distribution de la variable Churn :

In [51]:
#effectif tableau 
df["Churn"].value_counts()

Churn
False    2850
True      483
Name: count, dtype: int64

2850 abonnés sur 3333 sont fidèles. Pour otenir les fréquences, passer normalize=True à la fonction value_counts.

In [52]:
df["Churn"].value_counts(normalize=True)

Churn
False    0.855086
True     0.144914
Name: proportion, dtype: float64

### 4 Tri sur un tableau

Un DataFrame peut être trié par la valeur de l'une des variables (c'est-à-dire les colonnes). Par exemple, nous pouvons trier par Charge journalière totale (utilisez ascending=False pour trier par ordre décroissant) :

In [53]:
df.sort_values(by= "Total day charge", ascending= False).head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
365,CO,154,415,No,No,0,350.8,75,59.64,216.5,94,18.40,253.9,100,11.43,10.1,9,2.73,1,True
985,NY,64,415,Yes,No,0,346.8,55,58.96,249.5,79,21.21,275.4,102,12.39,13.3,9,3.59,1,True
2594,OH,115,510,Yes,No,0,345.3,81,58.70,203.4,106,17.29,217.5,107,9.79,11.8,8,3.19,1,True
156,OH,83,415,No,No,0,337.4,120,57.36,227.4,116,19.33,153.9,114,6.93,15.8,7,4.27,0,True
605,MO,112,415,No,No,0,335.5,77,57.04,212.5,109,18.06,265.0,132,11.93,12.7,8,3.43,2,True


Nous pouvons également trier suivant plusieurs colonnes :

In [54]:
df.sort_values(by= ["Churn", "Total day charge"], ascending= [True,False]).head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
688,MN,13,510,No,Yes,21,315.6,105,53.65,208.9,71,17.76,260.1,123,11.70,12.1,3,3.27,3,False
2259,NC,210,415,No,Yes,31,313.8,87,53.35,147.7,103,12.55,192.7,97,8.67,10.1,7,2.73,3,False
534,LA,67,510,No,No,0,310.4,97,52.77,66.5,123,5.65,246.5,99,11.09,9.2,10,2.48,4,False
575,SD,114,415,No,Yes,36,309.9,90,52.68,200.3,89,17.03,183.5,105,8.26,14.2,2,3.83,1,False
2858,AL,141,510,No,Yes,28,308.0,123,52.36,247.8,128,21.06,152.9,103,6.88,7.4,3,2.00,1,False


### 5 Indexing et récupération de données

Un DataFrame peut être indexé de différentes manières.

Pour récupérer une variable, vous pouvez utiliser : NomDataFrame['NomVariale']. 

Utilisons cette méthode pour répondre à une question : 

##### Quelle est la proportion d'abonnés ayant résilié dans notre base de données ?

In [55]:
# we use mean() because frequence is moyenne for bool
df["Churn"].mean()

np.float64(0.14491449144914492)

14,5 % est en fait assez mauvais pour une entreprise ; un tel taux de désabonnement peut entraîner la faillite de l'entreprise.

L'indexation booléenne avec une colonne est également très pratique. La syntaxe est df[P(df['Name'])], où P est une condition logique vérifiée pour chaque élément de la colonne Name. Le résultat d'une telle indexation est le DataFrame composé uniquement de lignes qui satisfont la condition P sur la colonne Name.

Profitons-en pour répondre à la question :

##### Combien de temps (en moyenne) les abonnés ayant résilié passent-ils au téléphone pendant la journée ?

In [56]:
df[ df["Churn"]== 1 ]["Total day charge"].mean()
# moyenne des charges du jour des abonnés résiliés

np.float64(35.17592132505176)

##### Quelle est la durée maximale des appels internationaux chez les utilisateurs fidèles (churn == 0) qui n'ont pas de forfait international ?

In [57]:
df[ ( df["Churn"] == 0 ) & ( df["International plan"] == "No") ] ["Total intl minutes"].max()

18.9

Les DataFrames peuvent être indexés par nom de colonne (étiquette) ou nom de ligne (index) ou par le numéro de série d'une ligne. La méthode loc() est utilisée pour l'indexation par nom, tandis que iloc() est utilisée pour l'indexation par numéro.


##### On désire afficher les valeurs des lignes avec un indice de 0 à 5 (inclus) correspondants aux colonnes "State" à "Area code"(inclus). 

In [58]:
df.loc[0:5, "State":"Area code"]

,State,Account length,Area code
0,KS,128,415
1,OH,107,415
2,NJ,137,415
3,OH,84,408
4,OK,75,415
5,AL,118,510


##### On désire afficher les valeurs des cinq premières lignes correspondant aux trois premières colonnes" 

In [59]:
#df[ df["Account length" >100].iloc[0:5, 0:3]

df.iloc[0:5, 0:3]

,State,Account length,Area code
0,KS,128,415
1,OH,107,415
2,NJ,137,415
3,OH,84,408
4,OK,75,415


##### Afficher la 1ère ligne du tableau

In [60]:
df.iloc[0:1]

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
0,KS,128,415,No,Yes,25,265.1,110,45.07,197.4,99,16.78,244.7,91,11.01,10.0,3,2.7,1,False


##### Afficher la dernière ligne du tableau

In [61]:
df.iloc[ -1:]

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
3332,TN,74,415,No,Yes,25,234.4,113,39.85,265.9,82,22.6,241.4,77,10.86,13.7,4,3.7,0,False


##### Application de fonctions aux cellules, aux colonnes et aux lignes

Pour appliquer des fonctions à chaque colonne, utilisez apply() :

In [62]:
# obtenir la valeur maximale pour chaque variable
import numpy as np
df.apply(np.max)

State                        WY
Account length              243
Area code                   510
International plan          Yes
Voice mail plan             Yes
Number vmail messages        51
Total day minutes         350.8
Total day calls             165
Total day charge          59.64
Total eve minutes         363.7
Total eve calls             170
Total eve charge          30.91
Total night minutes       395.0
Total night calls           175
Total night charge        17.77
Total intl minutes         20.0
Total intl calls             20
Total intl charge           5.4
Customer service calls        9
Churn                      True
dtype: object

- La méthode apply peut également être utilisée pour appliquer une fonction à chaque ligne. Pour ce faire, spécifiez axis =1. 


- Les fonctions Lambda sont très pratiques dans de tels scénarios. Par exemple, si nous devons sélectionner tous les états commençant par W, nous pouvons le faire comme ceci :

In [63]:
# afficher les State commençant par W
df[ df["State"].apply(lambda x : x [0] == "W") ].head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
9,WV,141,415,Yes,Yes,37,258.6,84,43.96,222.0,111,18.87,326.4,97,14.69,11.2,5,3.02,0,False
26,WY,57,408,No,Yes,39,213.0,115,36.21,191.1,112,16.24,182.7,115,8.22,9.5,3,2.57,0,False
44,WI,64,510,No,No,0,154.0,67,26.18,225.8,118,19.19,265.3,86,11.94,3.5,3,0.95,1,False
49,WY,97,415,No,Yes,24,133.2,135,22.64,217.2,58,18.46,70.6,79,3.18,11.0,3,2.97,1,False
54,WY,87,415,No,No,0,151.0,83,25.67,219.7,116,18.67,203.9,127,9.18,9.7,3,2.62,5,True


- La méthode map peut être utilisée pour remplacer des valeurs dans une colonne en passant un dictionnaire de la forme {ancienne_valeur : nouvelle_valeur} comme argument :

In [64]:
dico= {"No" : False, "Yes" : True}
df["International plan"]= df["International plan"].map(dico)
df.head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
0,KS,128,415,False,Yes,25,265.1,110,45.07,197.4,99,16.78,244.7,91,11.01,10.0,3,2.70,1,False
1,OH,107,415,False,Yes,26,161.6,123,27.47,195.5,103,16.62,254.4,103,11.45,13.7,3,3.70,1,False
2,NJ,137,415,False,No,0,243.4,114,41.38,121.2,110,10.30,162.6,104,7.32,12.2,5,3.29,0,False
3,OH,84,408,True,No,0,299.4,71,50.90,61.9,88,5.26,196.9,89,8.86,6.6,7,1.78,2,False
4,OK,75,415,True,No,0,166.7,113,28.34,148.3,122,12.61,186.9,121,8.41,10.1,3,2.73,3,False


- La même chose peut être faite avec la méthode replace :

In [65]:
df= df.replace({"Voice mail plan": dico})
df.head()

C:\Users\HP\AppData\Local\Temp\ipykernel_16936\466250223.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df= df.replace({"Voice mail plan": dico})


,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
0,KS,128,415,False,True,25,265.1,110,45.07,197.4,99,16.78,244.7,91,11.01,10.0,3,2.70,1,False
1,OH,107,415,False,True,26,161.6,123,27.47,195.5,103,16.62,254.4,103,11.45,13.7,3,3.70,1,False
2,NJ,137,415,False,False,0,243.4,114,41.38,121.2,110,10.30,162.6,104,7.32,12.2,5,3.29,0,False
3,OH,84,408,True,False,0,299.4,71,50.90,61.9,88,5.26,196.9,89,8.86,6.6,7,1.78,2,False
4,OK,75,415,True,False,0,166.7,113,28.34,148.3,122,12.61,186.9,121,8.41,10.1,3,2.73,3,False


### 6 Regroupement

En général, le regroupement des données dans Pandas fonctionne comme suit :

df.groupby(by=colonnes_de_regroupement)[colonnes_a_afficher].fonction()

- Tout d'abord, la méthode groupby divise les colonnes_de_regroupement par leurs valeurs. Ils deviennent un nouvel index dans le dataframe résultant.


- Ensuite, les colonnes d'intérêt sont sélectionnées (colonnes_a_afficher). 


- Si colonnes_a_afficher n'est pas inclus, toutes les clauses non groupby seront incluses.
     
     
- Enfin, une ou plusieurs fonctions sont appliquées aux groupes obtenus par colonnes sélectionnées.

##### Voici un exemple où nous regroupons les données en fonction des valeurs de la variable Churn et affichons les statistiques de trois colonnes dans chaque groupe :

In [67]:
# Exemple 1 : Fonction 'moyenne' pour la colonne de regroupement

colonnes= ["Total day minutes", "Total eve minutes"]
df.groupby( ["Churn"] )[colonnes].mean()

,Total day minutes,Total eve minutes
Churn,,
False,175.175754,199.043298
True,206.914079,212.410145


In [68]:
# Exemple 2 : statistiques descriptives pour la colonne de regroupement
df.groupby("Churn").agg({"Total intl minutes": "mean"})
# agg permet de calculer plusieurs statistiques descriptives sur une ou plusieurs 
# colonnes du DataFrame, en fonction d'une ou plusieurs colonnes de regroupement.

,Total intl minutes
Churn,
False,10.158877
True,10.700000


- Interprétation : les abonnés résiliés ont en moyenne consommé plus de minutes internationales que les abonnés non résiliés.

In [69]:
# Exemple 3 : Faisons la même chose, mais légèrement différemment en passant une liste de fonctions à agg() :

df.groupby("Churn").agg({"Total intl minutes": ["mean", "std"]})

Total intl minutes          
                    mean       std
Churn                             
False          10.158877  2.784489
True           10.700000  2.793190

- Interprétation : les abonnés résiliés ont en moyenne consommé plus de minutes internationales que les abonnés non résiliés, et les abonnés résiliés ont une plus grande variabilité dans leur consommation de minutes internationales que les abonnés non résiliés.

### 7 Tableaux croisés

Supposons que nous souhaitions connaître les effectifs ou les fréquences des modalités de la variable Churn selon la variable International Plan

In [70]:
# effectifs
pd.crosstab( df["Churn"], df["International plan"] )

International plan,False,True
Churn,,
False,2664,186
True,346,137


In [71]:
# fréquences
pd.crosstab( df["Churn"], df["International plan"], normalize=True )

International plan,False,True
Churn,,
False,0.79928,0.055806
True,0.10381,0.041104


On constate que la plupart des abonnés sont fidèles et n'utilisent pas de services supplémentaires (Forfait international/Messagerie vocale).

### 8 Tableaux croisés dynamiques 

Ils sont implémentés dans Pandas avec la méthode pivot_table avec les paramètres suivants :

- values : une liste de variables pour lesquelles calculer des statistiques,
     
- index : une liste de variables pour regrouper les données,

- aggfunc : quelles statistiques nous devons calculer pour les groupes, ex. somme, moyenne, maximum, minimum ou autre chose.

##### Exemple 1 : affichons le nombre total d'abonnés qui effectuent des appels le jour, le soir et la nuit par indicatif régional :

In [72]:
df.pivot_table( ["Total day calls", "Total eve calls", "Total night calls"], 
                ["Area code"], 
                aggfunc= "sum" )

,Total day calls,Total eve calls,Total night calls
Area code,,,
408,84216,83623,82995
415,166454,166334,166159
510,84082,83724,84505


##### Exemple 2 : affichons le nombre moyen journalier d'abonnés qui effectuent des appels le jour, le soir et la nuit par indicatif régional :


In [73]:
df.pivot_table(
     values=["Total day calls", "Total eve calls", "Total night calls"],
     index="Area code",
     aggfunc="mean"
)

,Total day calls,Total eve calls,Total night calls
Area code,,,
408,100.496420,99.788783,99.039379
415,100.576435,100.503927,100.398187
510,100.097619,99.671429,100.601190


### 9 Transformations de DataFrame

L'ajout de colonnes à un DataFrame est possible.

Par exemple, si nous voulons calculer le nombre total d'appels effectués par tous les abonnés, créons la série total_calls et collons-la au DataFrame :

In [79]:
# # Etape 1 : Création de la série total_calls
total_calls = (
     df["Total day calls"]
     + df["Total eve calls"]
     + df["Total night calls"]
     + df["Total intl calls"]
)

total_calls.head()

0    303
1    332
2    333
3    255
4    359
dtype: int64

In [80]:
# Etape 2 : Ajout de la colonne "Total calls" à la fin du DataFrame

# le paramètre loc est le nombre de colonnes après lesquelles nous insérons les nouvelles variables
# nous lui attribuons la valeur len(df.columns) pour le coller à la fin du dataframe

df.insert(loc=len(df.columns), column="Total calls", value=total_calls)

df.head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,...,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn,Total calls
0,KS,128,415,False,True,25,265.1,110,45.07,197.4,...,16.78,244.7,91,11.01,10.0,3,2.70,1,False,303
1,OH,107,415,False,True,26,161.6,123,27.47,195.5,...,16.62,254.4,103,11.45,13.7,3,3.70,1,False,332
2,NJ,137,415,False,False,0,243.4,114,41.38,121.2,...,10.30,162.6,104,7.32,12.2,5,3.29,0,False,333
3,OH,84,408,True,False,0,299.4,71,50.90,61.9,...,5.26,196.9,89,8.86,6.6,7,1.78,2,False,255
4,OK,75,415,True,False,0,166.7,113,28.34,148.3,...,12.61,186.9,121,8.41,10.1,3,2.73,3,False,359


- Il est possible d'ajouter directement une colonne au tableau sans créer la variable au préalable :

In [81]:
# Ajouter directement au DataFrame la somme des charges
df["Total charge"] = (
     df["Total day charge"]
     + df["Total eve charge"]
     + df["Total night charge"]
     + df["Total intl charge"]
)

df.head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,...,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn,Total calls,Total charge
0,KS,128,415,False,True,25,265.1,110,45.07,197.4,...,244.7,91,11.01,10.0,3,2.70,1,False,303,75.56
1,OH,107,415,False,True,26,161.6,123,27.47,195.5,...,254.4,103,11.45,13.7,3,3.70,1,False,332,59.24
2,NJ,137,415,False,False,0,243.4,114,41.38,121.2,...,162.6,104,7.32,12.2,5,3.29,0,False,333,62.29
3,OH,84,408,True,False,0,299.4,71,50.90,61.9,...,196.9,89,8.86,6.6,7,1.78,2,False,255,66.80
4,OK,75,415,True,False,0,166.7,113,28.34,148.3,...,186.9,121,8.41,10.1,3,2.73,3,False,359,52.09


### 10 Suppression de lignes ou de colonnes

Pour supprimer des colonnes ou des lignes :

- utilisez la méthode drop, en passant les index requis et le paramètre d'axe (1 si vous supprimez des colonnes, et rien ou 0 si vous supprimez des lignes). 


- L'argument inplace indique s'il faut modifier le DataFrame d'origine. 


- Avec inplace=False, la méthode drop ne modifie pas le DataFrame existant et en renvoie un nouveau avec des lignes ou des colonnes supprimées. 


- Avec inplace=True, cela modifie le DataFrame.

In [82]:
# suppression des colonnes récemment créées ("Total charge", "Total calls")
df.drop(columns=["Total charge", "Total calls"], inplace=True)
df.head(5)


,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
0,KS,128,415,False,True,25,265.1,110,45.07,197.4,99,16.78,244.7,91,11.01,10.0,3,2.70,1,False
1,OH,107,415,False,True,26,161.6,123,27.47,195.5,103,16.62,254.4,103,11.45,13.7,3,3.70,1,False
2,NJ,137,415,False,False,0,243.4,114,41.38,121.2,110,10.30,162.6,104,7.32,12.2,5,3.29,0,False
3,OH,84,408,True,False,0,299.4,71,50.90,61.9,88,5.26,196.9,89,8.86,6.6,7,1.78,2,False
4,OK,75,415,True,False,0,166.7,113,28.34,148.3,122,12.61,186.9,121,8.41,10.1,3,2.73,3,False


In [83]:
# suppression des lignes ayant les indices de position 1 et 2
df.drop(index=[1, 2], inplace=True)
df.head(5)

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
0,KS,128,415,False,True,25,265.1,110,45.07,197.4,99,16.78,244.7,91,11.01,10.0,3,2.70,1,False
3,OH,84,408,True,False,0,299.4,71,50.90,61.9,88,5.26,196.9,89,8.86,6.6,7,1.78,2,False
4,OK,75,415,True,False,0,166.7,113,28.34,148.3,122,12.61,186.9,121,8.41,10.1,3,2.73,3,False
5,AL,118,510,True,False,0,223.4,98,37.98,220.6,101,18.75,203.9,118,9.18,6.3,6,1.70,0,False
6,MA,121,510,False,True,24,218.2,88,37.09,348.5,108,29.62,212.6,118,9.57,7.5,7,2.03,3,False
